x-pack/plugins/infra/common/http_api/host_details/get_infra_services.ts

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  createLiteralValueFromUndefinedRT,
+  inRangeFromStringRt,
+  dateRt,
+  datemathStringRt,
+} from '@kbn/io-ts-utils';
+import * as rt from 'io-ts';
+
+export const sizeRT = rt.union([
+  inRangeFromStringRt(1, 100),
+  createLiteralValueFromUndefinedRT(10),
+]);
+export const assetDateRT = rt.union([dateRt, datemathStringRt]);
+
+export const servicesFiltersRT = rt.exact(
+  rt.type({
+    ['host.name']: rt.string,
+  })
+);
+
+export type ServicesFilter = rt.TypeOf<typeof servicesFiltersRT>;
+
+export const GetServicesRequestQueryRT = rt.intersection([
+  rt.strict({ from: assetDateRT, to: assetDateRT, filters: rt.string }),
+  rt.partial({
+    size: sizeRT,
+    validatedFilters: servicesFiltersRT,
+  }),
+]);
+
+export type GetServicesRequestQuery = rt.TypeOf<typeof GetServicesRequestQueryRT>;
+
+export interface ServicesAPIRequest {
+  filters: ServicesFilter;
+  from: string;
+  to: string;
+  size?: number;
+}
+
+const AgentNameRT = rt.union([rt.string, rt.null]);
+
+export const ServicesAPIQueryAggregationRT = rt.type({
+  services: rt.type({
+    buckets: rt.array(
+      rt.type({
+        key: rt.string,
+        latestAgent: rt.type({
+          top: rt.array(
+            rt.type({
+              sort: rt.array(rt.string),
+              metrics: rt.type({
+                'agent.name': AgentNameRT,
+              }),
+            })
+          ),
+        }),
+      })
+    ),
+  }),
+});
+
+export type ServicesAPIQueryAggregation = rt.TypeOf<typeof ServicesAPIQueryAggregationRT>;
+
+export const ServiceRT = rt.type({
+  'service.name': rt.string,
+  'agent.name': AgentNameRT,
+});
```

Comment:
I suggest not manually typing the ES responses. This it very brittle because as soon as the query is changed, the types will be invalid and need to be manually updated. You can get typed es client that returns typed responses like this:

https://github.com/elastic/kibana/blob/950c9345464be2cff136154c168b128b010b7318/x-pack/plugins/apm/server/routes/diagnostics/create_typed_es_client.ts#L16-L25

Reply:
This looks good and don't know of any good reason not do adopt this approach.  I'd like to focus on this [in a separate issue](https://github.com/elastic/kibana/issues/175268) and adopt it across our apis.  Thanks!